From 9b8a18843f93daa79325e4ff37535c2fcbad3af4 Mon Sep 17 00:00:00 2001
From: Yiannis Mandravellos <mandrav@gmail.com>
Date: Fri, 11 Aug 2023 09:02:38 +0300
Subject: [PATCH] lxd/network: fix bgp.ipv*.nexthop keys not being recognized

Signed-off-by: Yiannis Mandravellos <mandrav@gmail.com>
---
 lxd/network/driver_common.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lxd/network/driver_common.go b/lxd/network/driver_common.go
index 7b0a435c44fc..deeb1483312a 100644
--- a/lxd/network/driver_common.go
+++ b/lxd/network/driver_common.go
@@ -579,14 +579,14 @@ func (n *common) bgpValidationRules(config map[string]string) (map[string]func(v
 	rules := map[string]func(value string) error{}
 	for k := range config {
 		// BGP keys have the peer name in their name, extract the suffix.
-		if !strings.HasPrefix(k, "bgp.") {
+		if !strings.HasPrefix(k, "bgp.peers.") {
 			continue
 		}
 
 		// Validate remote name in key.
 		fields := strings.Split(k, ".")
 		if len(fields) != 4 {
-			return nil, fmt.Errorf("Invalid network configuration key: %s", k)
+			return nil, fmt.Errorf("Invalid network configuration key: %q", k)
 		}
 
 		bgpKey := fields[3]
