From f9c3bdf53bb93a5a71b8e23159a1fd02c8f96868 Mon Sep 17 00:00:00 2001
From: Thomas Hipp <thomas.hipp@canonical.com>
Date: Mon, 19 Sep 2022 16:23:43 +0200
Subject: [PATCH] client: Handle nil args in CopyImage

Signed-off-by: Thomas Hipp <thomas.hipp@canonical.com>
---
 client/lxd_images.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/client/lxd_images.go b/client/lxd_images.go
index 9f2a62419e8c..499751402225 100644
--- a/client/lxd_images.go
+++ b/client/lxd_images.go
@@ -675,7 +675,7 @@ func (r *ProtocolLXD) CopyImage(source ImageServer, image api.Image, args *Image
 	}
 
 	// Handle profile list overrides.
-	if args.Profiles != nil {
+	if args != nil && args.Profiles != nil {
 		if !r.HasExtension("image_copy_profile") {
 			return nil, fmt.Errorf("The server is missing the required \"image_copy_profile\" API extension")
 		}
