From 76493f38d709caca169b6f8747e979a2e04c97cb Mon Sep 17 00:00:00 2001
From: Thomas Parrott <thomas.parrott@canonical.com>
Date: Tue, 30 May 2023 08:32:35 +0100
Subject: [PATCH] lxd/device/nic/ovn: Enable hotplug for VMs

Fixes #11746

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
---
 lxd/device/nic_ovn.go | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lxd/device/nic_ovn.go b/lxd/device/nic_ovn.go
index 4e7fcb56ce29..5358e856df8b 100644
--- a/lxd/device/nic_ovn.go
+++ b/lxd/device/nic_ovn.go
@@ -49,6 +49,11 @@ type nicOVN struct {
 	network ovnNet // Populated in validateConfig().
 }
 
+// CanHotPlug returns whether the device can be managed whilst the instance is running.
+func (d *nicOVN) CanHotPlug() bool {
+	return true
+}
+
 // CanMigrate returns whether the device can be migrated to any other cluster member.
 func (d *nicOVN) CanMigrate() bool {
 	return true
