commit 384b57b555e5f133c84334f4634248fcdb87cf92
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Tue Jan 23 07:16:56 2018 +0100

    Update error string printed when group doesn't exist
    
diff --git a/lxd/endpoints/local_test.go b/lxd/endpoints/local_test.go
index 57b34c19b..340961371 100644
--- a/lxd/endpoints/local_test.go
+++ b/lxd/endpoints/local_test.go
@@ -70,7 +70,7 @@ func TestEndpoints_LocalUnknownUnixGroup(t *testing.T) {
 	err := endpoints.Up(config)
 
 	assert.EqualError(
-		t, err, "local endpoint: cannot get group ID of 'xquibaz': group: unknown group xquibaz")
+		t, err, "local endpoint: cannot get group ID of 'xquibaz': user: lookup groupname xquibaz: no such process")
 }
 
 // If another LXD endpoints is already listening on the unix socket, an error is
