commit 3e8363134642a7806d75d689603bd6a5483ec495
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Mon Aug 7 22:48:40 2023 +0200

    doc: Disable version lookup and switching

diff --git a/doc/.sphinx/_templates/sidebar/variant-selector.html b/doc/.sphinx/_templates/sidebar/variant-selector.html
index 59ae9452f..7cd0fe723 100644
--- a/doc/.sphinx/_templates/sidebar/variant-selector.html
+++ b/doc/.sphinx/_templates/sidebar/variant-selector.html
@@ -3,7 +3,6 @@
   {{ super() }}
 {% else %}
 <div class="version_dropdown">
-  Doc version: <button onclick="dropdown()" class="version_select" id="current"></button>
-  <div id="all-versions" class="available_versions"></div>
+  Doc version: 5.15
 </div>
 {% endif %}
diff --git a/doc/conf.py b/doc/conf.py
index 34c993277..12e551c0f 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -54,7 +54,7 @@ html_last_updated_fmt = ""
 html_favicon = ".sphinx/_static/download/favicon.ico"
 html_static_path = ['.sphinx/_static']
 html_css_files = ['custom.css']
-html_js_files = ['header-nav.js','version-switcher.js']
+html_js_files = ['header-nav.js']
 html_extra_path = ['.sphinx/_extra']
 
 html_theme_options = {
