commit 3213531c6d6cd761d09654245ef6177fbc4b4979
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Thu Dec 28 13:39:05 2023 +0100

    doc: Remove downloads from sphinx build

diff --git a/doc/custom_conf.py b/doc/custom_conf.py
index 81ca95647..2960a02cc 100644
--- a/doc/custom_conf.py
+++ b/doc/custom_conf.py
@@ -3,7 +3,6 @@ import os
 import sys
 import subprocess
 import yaml
-from git import Repo
 import filecmp
 
 # Custom configuration for the Sphinx documentation builder.
@@ -264,19 +263,6 @@ intersphinx_mapping = {
 
 html_extra_path = ['.sphinx/_extra']
 
-# Download and link swagger-ui files
-if not os.path.isdir('.sphinx/deps/swagger-ui'):
-    Repo.clone_from('https://github.com/swagger-api/swagger-ui', '.sphinx/deps/swagger-ui', depth=1, single_branch=True, b='v5.11.7')
-
-os.makedirs('.sphinx/_static/swagger-ui/', exist_ok=True)
-
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-bundle.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-bundle.js', '.sphinx/_static/swagger-ui/swagger-ui-bundle.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-standalone-preset.js', '.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui.css'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui.css', '.sphinx/_static/swagger-ui/swagger-ui.css')
-
 ### MAN PAGES ###
 
 # Find path to lxc client (different for local builds and on RTD)
diff --git a/doc/myconf.py b/doc/myconf.py
index 6707ddb81..939adc548 100644
--- a/doc/myconf.py
+++ b/doc/myconf.py
@@ -5,23 +5,8 @@ import stat
 import subprocess
 import tempfile
 import yaml
-from git import Repo
-import wget
 import filecmp
 
-# Download and link swagger-ui files
-if not os.path.isdir('.sphinx/deps/swagger-ui'):
-    Repo.clone_from('https://github.com/swagger-api/swagger-ui', '.sphinx/deps/swagger-ui', depth=1)
-
-os.makedirs('.sphinx/_static/swagger-ui/', exist_ok=True)
-
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-bundle.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-bundle.js', '.sphinx/_static/swagger-ui/swagger-ui-bundle.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-standalone-preset.js', '.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui.css'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui.css', '.sphinx/_static/swagger-ui/swagger-ui.css')
-
 ### MAN PAGES ###
 
 # Find path to lxc client (different for local builds and on RTD)
