commit 1cb5b84d65128dc55f27edd8fc64141aca76fe75
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Sun Aug 27 19:17:57 2023 +0200

    doc: Remove downloads from sphinx build

diff --git a/doc/conf.py b/doc/conf.py
index dceb3b49d..b72aa902b 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -5,21 +5,6 @@ import subprocess
 import sys
 import tempfile
 import yaml
-from git import Repo
-import wget
-
-# Download and link swagger-ui files
-if not os.path.isdir('.sphinx/deps/swagger-ui'):
-    Repo.clone_from('https://github.com/swagger-api/swagger-ui', '.sphinx/deps/swagger-ui', depth=1)
-
-os.makedirs('.sphinx/_static/swagger-ui/', exist_ok=True)
-
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-bundle.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-bundle.js', '.sphinx/_static/swagger-ui/swagger-ui-bundle.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-standalone-preset.js', '.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui.css'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui.css', '.sphinx/_static/swagger-ui/swagger-ui.css')
 
 # Project config.
 project = "LXD"
