commit 0ff7ab0e299f033026e39c059b654d9698ac9396
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Sun Aug 27 19:17:57 2023 +0200

    doc: Remove downloads from sphinx build

diff --git a/doc/conf.py b/doc/conf.py
index 6707ddb81..939adc548 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -5,23 +5,8 @@ import stat
 import subprocess
 import tempfile
 import yaml
-from git import Repo
-import wget
 import filecmp
 
-# Download and link swagger-ui files
-if not os.path.isdir('.sphinx/deps/swagger-ui'):
-    Repo.clone_from('https://github.com/swagger-api/swagger-ui', '.sphinx/deps/swagger-ui', depth=1)
-
-os.makedirs('.sphinx/_static/swagger-ui/', exist_ok=True)
-
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-bundle.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-bundle.js', '.sphinx/_static/swagger-ui/swagger-ui-bundle.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-standalone-preset.js', '.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui.css'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui.css', '.sphinx/_static/swagger-ui/swagger-ui.css')
-
 ### MAN PAGES ###
 
 # Find path to lxc client (different for local builds and on RTD)
