From 154b300535dc9f679fe6bceb60075bcbc854d984 Mon Sep 17 00:00:00 2001
From: Din Music <din.music@canonical.com>
Date: Wed, 25 Oct 2023 12:36:19 +0200
Subject: [PATCH] lxd/cluster/config: Add missing bool default values

Signed-off-by: Din Music <din.music@canonical.com>
---
 lxd/cluster/config/config.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lxd/cluster/config/config.go b/lxd/cluster/config/config.go
index 8a00bf595eb1..2436be209571 100644
--- a/lxd/cluster/config/config.go
+++ b/lxd/cluster/config/config.go
@@ -456,7 +456,7 @@ var ConfigSchema = config.Schema{
 	//  type: bool
 	//  scope: global
 	//  shortdesc: Whether to set `Access-Control-Allow-Credentials`
-	"core.https_allowed_credentials": {Type: config.Bool},
+	"core.https_allowed_credentials": {Type: config.Bool, Default: "false"},
 
 	// lxdmeta:generate(entity=server, group=core, key=core.https_trusted_proxy)
 	// Specify a comma-separated list of IP addresses of trusted servers that provide the client's address through the proxy connection header.
@@ -524,7 +524,7 @@ var ConfigSchema = config.Schema{
 	//  type: bool
 	//  scope: global
 	//  shortdesc: Whether to automatically trust clients signed by the CA
-	"core.trust_ca_certificates": {Type: config.Bool},
+	"core.trust_ca_certificates": {Type: config.Bool, Default: "false"},
 
 	// lxdmeta:generate(entity=server, group=candid-and-rbac, key=candid.api_key)
 	//
