From 93e1ed8bd354f1e3e2d1781cb70e2afe453d69a1 Mon Sep 17 00:00:00 2001
From: Thomas Parrott <thomas.parrott@canonical.com>
Date: Mon, 5 Dec 2022 09:40:33 +0000
Subject: [PATCH] lxd/resources/network: Ignore errors from ethtoolAddPortInfo

Fixes #11131 #11191

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
---
 lxd/resources/network.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lxd/resources/network.go b/lxd/resources/network.go
index 2473d9dcbadb..a127b1aed36c 100644
--- a/lxd/resources/network.go
+++ b/lxd/resources/network.go
@@ -248,16 +248,16 @@ func networkAddDeviceInfo(devicePath string, pciDB *pcidb.PCIDB, uname unix.Utsn
 				info.Infiniband = infiniband
 			}
 
-			// Attempt to add ethtool details (ignore failures)
 			if sysfsExists(filepath.Join(devicePath, "physfn")) {
 				// Getting physical port info for VFs makes no sense
 				card.Ports = append(card.Ports, *info)
 				continue
 			}
 
+			// Attempt to add ethtool details (ignore failures)
 			err = ethtoolAddPortInfo(info)
 			if err != nil {
-				return fmt.Errorf("Failed to add port info: %w", err)
+				continue
 			}
 
 			card.Ports = append(card.Ports, *info)
