commit f2239fb81a9d3464408afc1738ee554a85d39258
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Sun Aug 6 18:18:43 2023 +0200

    doc: Remove downloads from sphinx build

diff --git a/doc/conf.py b/doc/conf.py
index 0a7ec239a..34c993277 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -2,31 +2,6 @@ import datetime
 import os
 import sys
 import yaml
-from git import Repo
-import wget
-
-# Download and link swagger-ui files
-if not os.path.isdir('.sphinx/deps/swagger-ui'):
-    Repo.clone_from('https://github.com/swagger-api/swagger-ui', '.sphinx/deps/swagger-ui', depth=1)
-
-os.makedirs('.sphinx/_static/swagger-ui/', exist_ok=True)
-
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-bundle.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-bundle.js', '.sphinx/_static/swagger-ui/swagger-ui-bundle.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui-standalone-preset.js', '.sphinx/_static/swagger-ui/swagger-ui-standalone-preset.js')
-if not os.path.islink('.sphinx/_static/swagger-ui/swagger-ui.css'):
-    os.symlink('../../deps/swagger-ui/dist/swagger-ui.css', '.sphinx/_static/swagger-ui/swagger-ui.css')
-
-# Download and link images
-os.makedirs('.sphinx/_static/download/', exist_ok=True)
-
-if not os.path.isfile('.sphinx/_static/download/favicon.ico'):
-    wget.download("https://linuxcontainers.org/static/img/favicon.ico", ".sphinx/_static/download/favicon.ico")
-if not os.path.isfile('.sphinx/_static/download/containers.png'):
-    wget.download("https://linuxcontainers.org/static/img/containers.png", ".sphinx/_static/download/containers.png")
-if not os.path.isfile('doc/.sphinx/_static/download/containers.small.png'):
-    wget.download("https://linuxcontainers.org/static/img/containers.small.png", ".sphinx/_static/download/containers.small.png")
 
 # Project config.
 project = "LXD"
