From 33669481da63ed8ffecd3dad8d57f1024287d3b7 Mon Sep 17 00:00:00 2001
From: Thomas Parrott <thomas.parrott@canonical.com>
Date: Wed, 18 Jan 2023 09:35:17 +0000
Subject: [PATCH] shared/util: Update RunCommandCLocale to pass LANGUAGE=en env
 var

Fixes #11294

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
---
 shared/util.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/shared/util.go b/shared/util.go
index 061c15af441c..e1e0db799ef1 100644
--- a/shared/util.go
+++ b/shared/util.go
@@ -999,11 +999,11 @@ func RunCommandInheritFds(ctx context.Context, filesInherit []*os.File, name str
 	return stdout, err
 }
 
-// RunCommandCLocale runs a command with a LANG=C.UTF-8 environment set with optional arguments and
+// RunCommandCLocale runs a command with a LANG=C.UTF-8 and LANGUAGE=en environment set with optional arguments and
 // returns stdout. If the command fails to start or returns a non-zero exit code then an error is
 // returned containing the output of stderr.
 func RunCommandCLocale(name string, arg ...string) (string, error) {
-	stdout, _, err := RunCommandSplit(context.TODO(), append(os.Environ(), "LANG=C.UTF-8"), nil, name, arg...)
+	stdout, _, err := RunCommandSplit(context.TODO(), append(os.Environ(), "LANG=C.UTF-8", "LANGUAGE=en"), nil, name, arg...)
 	return stdout, err
 }
 
