From d8e611ada56e4546d568f887ec1fa8c751ae8ec5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Mathieu=20Border=C3=A9?= <mathieu.bordere@canonical.com>
Date: Sat, 5 Feb 2022 16:32:27 +0100
Subject: [PATCH] test/integration/uv: skip init/oom test on tmpfs

---
 test/integration/test_uv_init.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/test/integration/test_uv_init.c b/test/integration/test_uv_init.c
index 75045d46a..3d7cabc8c 100644
--- a/test/integration/test_uv_init.c
+++ b/test/integration/test_uv_init.c
@@ -3,6 +3,9 @@
 #include "../lib/runner.h"
 #include "../lib/uv.h"
 
+#include <linux/magic.h>
+#include <sys/vfs.h>
+
 /******************************************************************************
  *
  * Fixture with a non-initialized raft_io instance and uv dependencies.
@@ -144,10 +147,14 @@ static MunitParameterEnum oomParams[] = {
 TEST(init, oom, setUp, tearDown, 0, oomParams)
 {
     struct fixture *f = data;
-#if defined(__i686__)
     /* XXX: tmpfs seems to not support O_DIRECT */
-    return MUNIT_SKIP;
-#endif
+    struct statfs info;
+    int rv;
+    rv = statfs(f->dir, &info);
+    munit_assert_int(rv, ==, 0);
+    if (info.f_type == TMPFS_MAGIC) {
+        return MUNIT_SKIP;
+    }
 #if defined(__powerpc64__)
     /* XXX: fails on ppc64el */
     return MUNIT_SKIP;
