commit 9a66522b93d565d2ad950de202f18551f81d7c25
Author: Reto Gantenbein <reto.gantenbein@linuxmonk.ch>
Date:   Wed Apr 4 07:48:55 2018 +0200

    Disable some test that need access to GPG keyservers

diff --git a/shared/util_test.go b/shared/util_test.go
index 994eae3..24c09fa 100644
--- a/shared/util_test.go
+++ b/shared/util_test.go
@@ -5,15 +5,16 @@ import (
 	"io"
 	"log"
 	"os"
-	"path"
-	"path/filepath"
+/*	"path"
+	"path/filepath" */
 	"testing"
 
-	lxd "github.com/lxc/lxd/shared"
+//	lxd "github.com/lxc/lxd/shared"
 	"github.com/stretchr/testify/require"
 	"gopkg.in/flosch/pongo2.v3"
 )
 
+/* Disable online tests
 func TestVerifyFile(t *testing.T) {
 	wd, err := os.Getwd()
 	if err != nil {
@@ -111,6 +112,7 @@ func TestCreateGPGKeyring(t *testing.T) {
 	require.False(t, lxd.PathExists(keyring), "File should not exist")
 	os.RemoveAll(path.Dir(keyring))
 }
+*/
 
 func TestRenderTemplate(t *testing.T) {
 	tests := []struct {
